engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java

```diff
@@ -0,0 +1,353 @@
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    private static Instant instantForOffset(LocalDate date, LocalTime time, ZoneOffset offset) {
+      return date.atTime(time).toInstant(offset);
+    }
+
+    @Specialization(guards = "iop.isNull(value)")
+    static void doNull(
+        ArrowFixedSizeArrayBuilder receiver,
+        long index,
+        Object value,
+        @Cached.Shared("interop") @CachedLibrary(limit = "1") InteropLibrary iop)
+        throws UnsupportedMessageException {
+      if (receiver.sealed) {
+        throw UnsupportedMessageException.create();
+      }
+      receiver.buffer.setNull((int) index);
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Int8")
+    public static void doByte(
+        ArrowFixedSizeArrayBuilder receiver,
+        long index,
+        Object value,
+        @Cached.Shared("interop") @CachedLibrary(limit = "1") InteropLibrary iop)
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      validAccess(receiver, index);
+      if (!iop.fitsInByte(value)) {
+        throw UnsupportedMessageException.create();
+      }
+      receiver.buffer.put(typeAdjustedIndex(index, receiver.unit), (iop.asByte(value)));
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Int16")
+    public static void doShort(
+        ArrowFixedSizeArrayBuilder receiver,
+        long index,
+        Object value,
+        @Cached.Shared("interop") @CachedLibrary(limit = "1") InteropLibrary iop)
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      validAccess(receiver, index);
+      if (!iop.fitsInShort(value)) {
+        throw UnsupportedMessageException.create();
+      }
+      receiver.buffer.putShort(typeAdjustedIndex(index, receiver.unit), (iop.asShort(value)));
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Int32")
+    public static void doInt(
+        ArrowFixedSizeArrayBuilder receiver,
+        long index,
+        Object value,
+        @Cached.Shared("interop") @CachedLibrary(limit = "1") InteropLibrary iop)
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      validAccess(receiver, index);
+      if (!iop.fitsInInt(value)) {
+        throw UnsupportedMessageException.create();
+      }
+      receiver.buffer.putInt(typeAdjustedIndex(index, receiver.unit), (iop.asInt(value)));
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Int64")
+    public static void doLong(
+        ArrowFixedSizeArrayBuilder receiver,
+        long index,
+        Object value,
+        @Cached.Shared("interop") @CachedLibrary(limit = "1") InteropLibrary iop)
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      validAccess(receiver, index);
+      if (!iop.fitsInLong(value)) {
+        throw UnsupportedMessageException.create();
+      }
+      receiver.buffer.putLong(typeAdjustedIndex(index, receiver.unit), (iop.asLong(value)));
+    }
```

Comment:
I don't see any `@Fallback` or catch-all `@Specialization`. What happens if you try to write wrong value type into wrong array type? Will it yield user friendly `UnsupportedTypeException` or will it crash on missing specialization?